contracts/OpConsensusStateManager.sol

```diff
@@ -0,0 +1,127 @@
+
+
+// OptimisticConsensusStateManager manages the optimistic consensus
+// states. It keeps track a configurable number of optimistic
+// consensus states. IBC module uses this class to access an
+// optimistic consensus state at a given height to perform membership
+// proof.
+library OptimisticConsensusStateManager {
+    enum OpConsensusStateStatus { NOT_APPLIED, PENDING, TRUSTED }
+
+    struct ManagedOpConsensusState {
+        OptimisticConsensusState state;
+        OpConsensusStateStatus status;
+    }
+
+    struct DataStore {
+        uint32 fraudProofWindowSeconds;
+        uint64 capacity;
+        uint64 size;
+        mapping(uint256 => ManagedOpConsensusState) consensusStates;
+        uint256[] heights;
+    }
+
+    // addOpConsensusState adds an OptimisticConsensusState to
+    // internal store. The newly added op consensus state has a
+    // PENDING status.
+    // 
+    // adding a duplicate OptimisticConsensusState is an idempotent
+    // operation. it doesn't change the fraud proof end time or change
+    // the state of an already trusted OptimisticConsensusState.
+    //
+    // it is an error to update a pending OptimisticConsensusState
+    // with different values. such operation must be done through the
+    // sumitting fraud proof API.
+    //
+    // the status and the fraud proof end time are returned.
+    function addOpConsensusState(DataStore storage self, OptimisticConsensusState calldata state)
+        external returns (OpConsensusStateStatus status, uint256 fraudProofEndTime) {
+        ManagedOpConsensusState storage s = self.consensusStates[state.height];
+
+        if (s.status == OpConsensusStateStatus.TRUSTED) {
+            return (OpConsensusStateStatus.TRUSTED, 0);
+        }
+
+        if (s.status == OpConsensusStateStatus.PENDING) {
+            uint256 fraudProofEnd = s.state.time + self.fraudProofWindowSeconds;
+            if (block.timestamp > fraudProofEnd) {
+                s.status = OpConsensusStateStatus.TRUSTED;
+                return (OpConsensusStateStatus.TRUSTED, 0);
+            }
+           
+            // duplicate state update is a noop.
+            if (isOptimisticConsensusStateEqual(s.state, state)) {
+                return (OpConsensusStateStatus.PENDING, s.state.time + self.fraudProofWindowSeconds);
+            }
+
+            revert('cannot update a pending optimistic consensus state with different values, please submit fraud proof instead');
+        }
+
+        // this is a new OptimisticConsensusState.
+        s.state = state;
+        s.status = OpConsensusStateStatus.PENDING;
+        // n.b, don't trust the timestamp sent by the caller.
+        s.state.time = block.timestamp;
+        self.heights.push(state.height);
+        self.size++;
+
+        // delete the op consensus state with the smallest
+        // height(i.e., oldest) from internal store.
+        uint256 smallestHeight = self.heights[0];
+        if (self.size > self.capacity) {
```

Comment:
Shift operations are incredibly expensive in the EVM. Every value that gets updated costs 20k gas.

A ZK update is ~330k gas, the more gas this optimistic update costs the less it even makes sense to invest the engineering effort into doing this. We want to optimize for minimal gas overhead.

Reply:
I'll profile the gas cost, but want to clarify what you meant by shifting here. We are finding the smallest element, and then swapping it with the last element, and then pop the last element. it's less expensive than a real shift.

Reply:
Oh woops didn't read properly. Yea this would be less expensive.